### DISCUSSION TITLE
Browse/filter for courses by spoken languge

### DISCUSSION THREAD

**Discussion by kurmasz** [dodona-edu/feedback-and-support#57] [Feature requests]
Is there a way to browse/filter courses by spoken language? (In particular, is there a way for me to browse only courses in English?)

**Comment by jorg-vr** [comment]
There is not currently. We currently don't even keep track of the language of a course.

But this does seem like a sensible thing to add, I've made an issue out of this: #6039 

In the meantime: are you looking for something specific?
I can manually recommend you some English courses

**Comment by kurmasz** [comment]
I am looking for more short, coding-bat-like practice problems for my Python students. (This is CodingBat, if you haven't run across it before: https://codingbat.com/python

**Comment by pdawyndt** [comment]
@kurmasz When searching for individual exercises (not for entire courses) you can filter by natural language. Once you have found an exercise of the type your are looking for, you can also filter based on the repository the exercise comes from (which probably will have exercises supporting the same natural language and level. Just click the name of the repo below the exercise to use the repo as a search filter:

![image](https://github.com/user-attachments/assets/e545d8e2-268f-4e68-8c09-26d827595d03)

**Comment by pdawyndt** [comment]
@kurmasz Also consider adding your own exercises to Dodona if you have your own requirements. Here you find the tutorial: https://docs.dodona.be/en/references/exercise-description/

I took the first CodingBat Python warmup exercise ([`sleep_in`](https://codingbat.com/prob/p173401)) as an example and added it to Dodona:

https://dodona.be/en/activities/1228610798/

Here are the specific steps you need to follow:

1. create a new [exercise repository](https://docs.dodona.be/en/guides/exercises/new-exercise-repo/) to contain all your exercises
   - [global structure of exercise repository](https://docs.dodona.be/en/references/repository-directory-structure/)
2. create a new programming exercise according to this [directory structure](https://docs.dodona.be/en/references/exercise-directory-structure/)
   - [add a description](https://docs.dodona.be/en/references/exercise-description/) in Markdown or HTML
   - [add a config file](https://docs.dodona.be/en/references/exercise-config/)
   - [add a test suite](https://docs.dodona.be/en/references/tested/)

For the sample exercise I took over the description and added some Markdown styling (file `description/description.en.md`)

```markdown
The parameter `weekday` is `True` if it is a weekday, and the parameter `vacation` is `True` 
if we are on vacation. We sleep in if it is not a weekday or we're on vacation. Return `True` 
if we sleep in.
```

I also added boilerplate code that initially fills the code editor (file `description/boilerplate/boilerplate.en`), containing the header line of the function definition and a docstring containing three sample function calls with the expected return value:

```python
def sleep_in(weekday, vacation):

    """
    >>> sleep_in(False, False)
    True
    >>> sleep_in(True, False)
    False
    >>> sleep_in(False, True)
    True
    """
```

This config file sets the title for the exercise, sets the programming language to Python and specifies the [**TESTed**](https://docs.dodona.be/en/references/tested/) judge is used to assess submitted solutions, with test specifications taken from the `evaluation/suite.yaml` file.

```json
{
  "description": {
    "names": {
      "en": "Sleep in",
      "nl": "Sleep in"
    }
  },
  "programming_language": "python",
  "evaluation": {
    "handler": "tested",
    "test_suite": "suite.yaml"
  }
}
```

The test suite itself looks like this (file `evaluation/suite.yaml`; tests the four possible argument combinations):

```yaml
- unit: "Sleep in"
  scripts:
    - expression: "sleep_in(False, False)"
      return: True
    - expression: "sleep_in(True, False)"
      return: False
    - expression: "sleep_in(False, True)"
      return: True
    - expression: "sleep_in(True, True)"
      return: True
```

As a reference for yourself and colleagues (not accessible for students) you can also add your sample solution to the exercise directory (file `solution/solution.py`):

```python
def sleep_in(weekday, vacation):

    """
    >>> sleep_in(False, False)
    True
    >>> sleep_in(True, False)
    False
    >>> sleep_in(False, True)
    True
    """

    return not weekday or vacation
```

That's it. That gives you a working exercise on Dodona, and provides you with all the freedom to tweak the exercise to your own needs. *May the source be with you*.

**Reply by pdawyndt** [reply]
Here's the entire exercise directory as a ZIP archive: [sleep in.zip](https://github.com/user-attachments/files/18292370/sleep.in.zip)

**Comment by bmesuere** [comment]
This has been implemented!
